### PR TITLE
txHandler: applications rate limiter

### PR DESCRIPTION
**Pull request by algorandskiy**
## Summary

Rate limit incoming apps based on app id + sender IP addr. It precedes enqueueing into backlog in `TxHandler.processIncomingTxn`. It takes effect only if enabled and the backlog is more than 1/2 full.
The implementation uses sharded map with sliding window limiter. The sliding window data is used to for least recently used eviction (list based).

There are two hashes used: 1) memhash64 for app to bucket mapping 2) blake2b for app id + sender caching.
Importantly, the implementation is trying to compromise on memory usage and shrinks 32 bytes blake2b to 8 bytes. It appears OK since blake2b output is uniform, and input data is salted making impossible to censor app + relay pairs.

As benchmark (below) shows there is almost no penalty (5%) on eviction when 94% of operations cause eviction.

## Test Plan

1. Unit tests: currently 98% but missing an edge case with eviction and reusing.
2. Benchmarks
```
BenchmarkAppRateLimiter/multi_bucket_no_evict-8         	  179841	      6572 ns/op	         0.4306 %_drop	    1340 B/op	      16 allocs/op
BenchmarkAppRateLimiter/single_bucket_no_evict-8        	  152894	      6859 ns/op	         0.6651 %_drop	    1364 B/op	      17 allocs/op
BenchmarkAppRateLimiter/single_bucket_w_evict-8         	  147360	      7174 ns/op	         0 %_drop	    1408 B/op	      19 allocs/op
--- BENCH: BenchmarkAppRateLimiter/single_bucket_w_evict-8
    appRateLimiter_test.go:336: # evictions 15392, time 2119 us
    appRateLimiter_test.go:336: # evictions 139168, time 20589 us
```
3. Real apps traffic
Used real recorded transactions and approximated traffic by reported connected peers metric during the last high traffic event on Sep 1st 12:01 pm - 3:28pm. Got about 6M transactions and about 10k unique key pairs (no truncated hash collisions).
Run few benchmarks by using tx messages from the generated data set and got the following data:
  * 4m11s to read the data set and feed into txHandler without rate limiting
  * 4m45s to read and apply rate limiter
  * using time.Now() gave 98% drop rate
  * using tx timestamp (with some uniform approximation) gave 92% drop rate with the same run time
  * using tx timestamp but with random order gave 1% drop rate with the same run time (different workloads, same runtime.

Visualized acceptance rate:
image

4. Profiling / Resources

|       | mem prof | cpu prof |
|:---:|:---------:|:---------:|
| orig impl |  |  |
| memory fixes | image | image |

Memory overhead: 
1. `sync.Pool` for keys/buckets gathering 0.5% alloc_space, 25% alloc_object on a heavy load.
2. original version: 2.9% alloc_space, 38% alloc_object. Most allocs from keys/buckets gathering and `go-deadlock.lock()` (fixed in https://github.com/algorand/go-deadlock/pull/2):
```
/go/pkg/mod/github.com/algorand/go-deadlock@v0.2.2/deadlock.go

  Total:      1.55GB     1.55GB (flat, cum)  0.54%
    119            .          .           		m.muId = currID 
    120            .          .           		currID++ 
    121            .          .           	} 
    122            .          .           	counterMu.Unlock() 
    123            .          .            
    124       1.55GB     1.55GB           	lock(m.mu.Lock, m, false) 
```

5. Scenario1s run (same cmd as https://github.com/algorand/go-algorand/pull/5760) => no perf degradation observed.
```
relay summary: 4893.13 TPS, 3.19s/block, tx 26.2MB/s, rx 17.4MB/s
```

### CODE REVIEW

File: data/appRateLimiter.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+func (r *appRateLimiter) shouldDropInner(txgroup []transactions.SignedTxn, origin []byte, now time.Time) bool {
+	var keys []crypto.Digest
+	// TODO: check memory allocs
+	var buf [8 + 4]byte // uint64 + [4]byte
+	for i := range txgroup {
+		if txgroup[i].Txn.Type == protocol.ApplicationCallTx {
+			binary.LittleEndian.PutUint64(buf[:8], uint64(txgroup[i].Txn.ApplicationID))
+			toBeHashed := append(buf[8:], origin...)
+			d := crypto.Digest(blake2b.Sum256(toBeHashed))
```
Comment (bbroder-uji) [inline data/appRateLimiter.go:96]:
```suggestion
			copy (buf[8:], origin)
			d := crypto.Digest(blake2b.Sum256(buf))
```

Comment (algorandskiy) [reply data/appRateLimiter.go:96]: Was not sure what how to get conn data from p2p so went with safe append option. Eventually it will depends on a underlying cache/map choice, if no bucketing then no need to hash, appid + ip concatenation might be good enough. Also not sure if there should be two levels: appid and ip to give more fairness to other apps.

Comment (bbroder-uji) [reply data/appRateLimiter.go:96]: Aren't you trying to make a key that is the hash of the 8 byte app id + the 4 IP bytes?
in your code the length of toBeHashed is 8.
toBeHashed := append(buf[8:], origin...)
buf[8:] is length 4, and origin is length 4.

File: data/appRateLimiter.go [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+			d := crypto.Digest(blake2b.Sum256(buf[:8+copied]))
+			// TODO: since blake2 is a crypto hash function it seems OK to shrink 32 bytes digest to
+			// 16 or 12 bytes.
+			// Rationale: we expect thousands of apps sent from thousands of peers,
+			// so required millions of unique pairs => 12 or 16 bytes should be enough.
+			keys = append(keys, d)
+			if len(txgroup[i].Txn.ForeignApps) > 0 {
+				for j := range txgroup[i].Txn.ForeignApps {
+					binary.LittleEndian.PutUint64(buf[:8], uint64(txgroup[i].Txn.ForeignApps[j]))
+					copied = copy(buf[8:], origin)
```
Comment (bbroder-uji) [inline data/appRateLimiter.go:105]: not needed?

File: data/appRateLimiter.go [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+		// TODO: evict 10% oldest entries?
+		var oldestKey crypto.Digest
+		var oldestInterval int64 = math.MaxInt64
+		for k, v := range r.buckets[b] {
+			interval := v.interval.Load()
+			if interval < oldestInterval {
+				oldestKey = k
+				oldestInterval = interval
+			}
+		}
```
Comment (bbroder-uji) [inline data/appRateLimiter.go:80]: Ring buffers holding the bucket keys in entry order could avoid this range search if you want to spend the x megs of memory on it.

Comment (algorandskiy) [reply data/appRateLimiter.go]: used LRU lists

Comment (bbroder-uji) [reply data/appRateLimiter.go]: i saw, nice work pavel! reading it now

File: data/appRateLimiter.go [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+// makeAppRateLimiter creates a new appRateLimiter from the parameters:
+// maxCacheSize is the maximum number of entries to keep in the cache to keep it memory bounded
+// maxAppPeerRate is the maximum number of admitted apps per peer per second
+// serviceRateWindow is the service window
+func makeAppRateLimiter(maxCacheSize uint64, maxAppPeerRate uint64, serviceRateWindow time.Duration) *appRateLimiter {
+	// convert target per app rate to per window service rate
+	serviceRatePerWindow := maxAppPeerRate * uint64(serviceRateWindow/time.Second)
+	maxBucketSize := maxCacheSize / numBuckets
+	r := &appRateLimiter{
+		maxBucketSize:        maxBucketSize,
```
Comment (bbroder-uji) [inline data/appRateLimiter.go:77]: if you do it here you can stop casting it later
```suggestion
		maxBucketSize:        int(maxBucketSize),
```

File: data/appRateLimiter.go [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+		} else {
+			entry.cur.Add(1)
+		}
+
+		curFraction := r.fraction(now)
+		rate := uint64(float64(entry.prev.Load())*(1-curFraction)) + entry.cur.Load()
+
+		if rate > r.serviceRatePerWindow {
+			return true
+		}
```
Comment (bbroder-uji) [inline data/appRateLimiter.go:179]: The entry modifications aren't rolled back when the rate is exceeded.  So the rate is more of an attempted app rate than an admitted app rate.  A more nimble txgroup that might have been accepted had it gone first is now blocked by the effects of the prior dropped txgroups.

File: data/appRateLimiter.go [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+	entry, ok := r.buckets[b][key]
+	if ok {
+		el := entry.lruElement
+		r.lrus[b].MoveToFront(el)
+	} else {
+		el := r.lrus[b].PushFront(key)
+		entry = &appRateLimiterEntry{lruElement: el}
+		entry.cur.Store(1)
+		entry.interval.Store(curInt)
+		r.buckets[b][key] = entry
```
Comment (bbroder-uji) [inline data/appRateLimiter.go:116]: Since the entry is created and destroyed only in this same locked method, maybe using a sync pool is an easy reach.

File: data/appRateLimiter.go [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+			appIdx := txgroup[i].Txn.ApplicationID
+			// hash appIdx into a bucket, do not use modulo since it could
+			// assign two vanilla (and presumable, popular) apps to the same bucket.
+			buckets = append(buckets, int(memhash64(uint64(appIdx), seed)%uint64(numBuckets)))
+			keys = append(keys, txnToDigest(appIdx))
+			if len(txgroup[i].Txn.ForeignApps) > 0 {
+				for _, appIdx := range txgroup[i].Txn.ForeignApps {
+					buckets = append(buckets, int(memhash64(uint64(appIdx), seed)%uint64(numBuckets)))
+					keys = append(keys, txnToDigest(appIdx))
+				}
```
Comment (bbroder-uji) [inline data/appRateLimiter.go:220]: maybe since you're doing lambdas anyway maybe you add 
```
txnToBucket := func(appIdx basics.AppIndex) int {
     return int(memhash64(uint64(appIdx), seed)%uint64(numBuckets)
}
```
then you can avoid the double-definition below

File: data/appRateLimiter.go (hunk 5 above)
Comment (algorandskiy) [reply data/appRateLimiter.go]: nice catch, fixed

File: data/appRateLimiter.go (hunk 4 above)
Comment (algorandskiy) [reply data/appRateLimiter.go:81]: done

File: data/appRateLimiter.go (hunk 7 above)
Comment (algorandskiy) [reply data/appRateLimiter.go]: done

File: data/appRateLimiter.go (hunk 6 above)
Comment (algorandskiy) [reply data/appRateLimiter.go]: sure, in my list after adding some final benchmarks to get a base line performance.

File: data/appRateLimiter.go [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+			}
+
+			// admitted, update the entry
+			if thisInterval {
+				entry.cur.Add(1)
+			} else {
+				entry.prev.Store(newPrev)
+				entry.cur.Store(1)
+				entry.interval.Store(curInt)
+			}
```
Comment (bbroder-uji) [inline data/appRateLimiter.go:192]: This rolls back the rate violating key but it still doesn't roll back the earlier admitted-now-rejected keys in the range. 
1. maybe prune only once at the end of a successful shouldDropKeys instead of on each newEntryLocked and avoid depruning in rollback and the double-check performed below. 
2. You'll have to keep track of map inserts performed in newEntryLocked and the initial value of each entry (taken before the optimistic unlock) to properly do the rollback.

Comment (algorandskiy) [reply data/appRateLimiter.go]: rewrote this by reverting this two locks implementation back to single lock. Incrementing after admission. Updating LRU on attempts.

**Comment by bbroder-uji** [comment]
Incrementing after admission is a smart way to keep the load capped at the max rate.  
Looking good!

File: data/appRateLimiter.go [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+			// admit and increment
+			entry.cur.Add(1)
+			continue
+		}
+
+		rate := int64(float64(entry.prev.Load())*(1-curFraction)) + entry.cur.Load() + 1
+		if rate > int64(r.serviceRatePerWindow) {
+			return true
+		}
+		entry.cur.Add(1)
```
Comment (algorandskiy) [inline data/appRateLimiter.go:194]: this portion can be done under a mutex lock as well, it is a bit slower (7 us per `shouldDropKeys` vs 6.7 us as current). The very first version had only map lookup and lru update under a lock, but it ended up to be very hairy code to maintain counters and `curInt`.

File: data/appRateLimiter.go [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+		maxBucketSize:        maxBucketSize,
+		serviceRatePerWindow: serviceRatePerWindow,
+		serviceRateWindow:    serviceRateWindow,
+		seed:                 crypto.RandUint64(),
+	}
+	crypto.RandBytes(r.salt[:])
+
+	for i := 0; i < numBuckets; i++ {
+		r.buckets[i] = make(map[keyType]*appRateLimiterEntry)
+		r.lrus[i] = util.NewList[keyType]().AllocateFreeNodes(maxBucketSize)
```
Comment (jannotti) [inline data/appRateLimiter.go:92]: I don't know the list implementation well, but I don't see why there would be an advantage to pre-allocating here.  Why pay the memory and cpu time now, when you might not ever get around to needing it?

File: data/appRateLimiter.go [hunk 11]
```diff
@@ -0,0 +1,267 @@
...
+// fraction calculates the fraction of the interval that is elapsed since the given time
+func (r *appRateLimiter) fraction(now time.Time) float64 {
+	return float64(now.UnixNano()%int64(r.serviceRateWindow)) / float64(r.serviceRateWindow)
+}
+
+// shouldDrop returns true if the given transaction group should be dropped based on the
+// on the rate for the applications in the group: the entire group is dropped if a single application
+// exceeds the rate.
+func (r *appRateLimiter) shouldDrop(txgroup []transactions.SignedTxn, origin []byte) bool {
+	return r.shouldDropInner(txgroup, origin, time.Now())
```
Comment (jannotti) [inline data/appRateLimiter.go:161]: Can we avoid the word "Inner", since we're talking about transaction and it's easy to wonder what "inner transactions" have to do with this.

How about `shouldDropAt()`?

File: data/appRateLimiter.go [hunk 12]
```diff
@@ -0,0 +1,267 @@
...
+	}
+	return r.shouldDropKeys(buckets, keys, now)
+}
+
+func (r *appRateLimiter) shouldDropKeys(buckets []int, keys []keyType, now time.Time) bool {
+	curInt := r.interval(now)
+	curFraction := r.fraction(now)
+
+	for i := range keys {
+		key := keys[i]
```
Comment (jannotti) [inline data/appRateLimiter.go:179]:
```suggestion
	for i, key := range keys {
```

File: data/txHandler.go [hunk 13]
```diff
@@ -640,6 +650,12 @@ func (handler *TxHandler) processIncomingTxn(rawmsg network.IncomingMessage) net
 		}
 	}
 
+	// rate limit per application in a group. Limiting any app in a group drops the entire message.
+	if handler.appLimiter != nil && handler.appLimiter.shouldDrop(unverifiedTxGroup, rawmsg.Sender.(network.IPAddressable).IPAddr()) {
```
Comment (jannotti) [inline data/txHandler.go:654]: We discussed offline, but should probably take only the top 64 bits of an ipv6 address.

File: data/appRateLimiter.go (hunk 9 above)
Comment (bbroder-uji) [reply data/appRateLimiter.go:194]: the only gotcha here is that without the lock the entry could be not in the list anymore at this point.

File: data/txHandler.go (hunk 13 above)
Comment (bbroder-uji) [reply data/txHandler.go:654]: i agree excellent catch

Comment (algorandskiy) [reply data/txHandler.go]: fixed, use either ipv4, or 4 lower bytes if embedded address, or high 8 bytes of ipv6

**Review by algorandskiy — COMMENTED** [review]
Thanks for the review, fixed all remarks, and modified the limiter to kick in only if backlog size >= half of its max capacity.

File: data/appRateLimiter.go [hunk 14]
```diff
@@ -241,17 +267,17 @@ func txgroupToKeys(txgroup []transactions.SignedTxn, origin []byte, seed uint64,
 			appIdx := txgroup[i].Txn.ApplicationID
 			// hash appIdx into a bucket, do not use modulo since it could
```
Comment (bbroder-uji) [inline data/appRateLimiter.go:268]:
```suggestion
			// hash appIdx into a bucket, do not use modulo without hashing first since it could
```

File: network/wsPeer.go [hunk 15]
```diff
@@ -391,6 +392,32 @@ func (wp *wsPeer) IPAddr() []byte {
...
+				return false
+			}
+		}
+		return true
+	}
+
+	ip := wp.IPAddr()
+	if len(ip) == 0 || len(ip) == net.IPv4len || len(ip) != net.IPv6len {
+		return ip
+	}
```
Comment (bbroder-uji) [inline network/wsPeer.go:413]: `len(ip) != net.IPV6len` takes care of the other two predicates

File: data/txHandler.go [hunk 16]
```diff
@@ -588,7 +588,9 @@ func (handler *TxHandler) processIncomingTxn(rawmsg network.IncomingMessage) net
 
 	var err error
 	var capguard *util.ErlCapacityGuard
+	var congested bool
 	if handler.erl != nil {
+		congested = float64(cap(handler.backlogQueue))*0.5 < float64(len(handler.backlogQueue))
```
Comment (bbroder-uji) [inline data/txHandler.go:593]: maybe 0.5 is a config?

File: data/appRateLimiter.go (hunk 14 above)
Comment (algorandskiy) [reply data/appRateLimiter.go:268]: done

File: data/txHandler.go (hunk 16 above)
Comment (algorandskiy) [reply data/txHandler.go:593]: done

File: network/wsPeer.go (hunk 15 above)
Comment (algorandskiy) [reply network/wsPeer.go:413]: done

File: data/txHandler.go [hunk 17]
```diff
@@ -640,6 +658,12 @@ func (handler *TxHandler) processIncomingTxn(rawmsg network.IncomingMessage) net
 		}
 	}
 
+	// rate limit per application in a group. Limiting any app in a group drops the entire message.
+	if handler.appLimiter != nil && congested && handler.appLimiter.shouldDropAt(unverifiedTxGroup, rawmsg.Sender.(network.IPAddressable).RoutingAddr(), rawmsg.Received) {
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply data/txHandler.go]: done, use time.Now()

**Comment by algorandskiy** [comment]
I rerun unlimited (~master) vs app limiter benchmarks (`TxBacklogAppTxPerSecondRate=100` and backlog drain rate 10k tx messages/sec):
```
BenchmarkAppRateLimiterTxHandlerLogs/unlimited
    appRateLimiter_test.go:941: backlog: admitted 750238 (5070072 dropped) in 1m24.351858833s
    appRateLimiter_test.go:942: erl dropped 0, app limiter dropped 0
BenchmarkAppRateLimiterTxHandlerLogs/unlimited-10         	       1	87664614417 ns/op	297839384760 B/op	329630204 allocs/op

BenchmarkAppRateLimiterTxHandlerLogs/app_only
    appRateLimiter_test.go:941: backlog: admitted 844550 (1348901 dropped) in 1m31.448261541s
    appRateLimiter_test.go:942: erl dropped 0, app limiter dropped 3624011
BenchmarkAppRateLimiterTxHandlerLogs/app_only-10          	       1	100746523250 ns/op	298025311832 B/op	326026307 allocs/op
```
with limiter ON some more txns admitted so it does not make thinks worse.

**Comment by algorandskiy** [comment]
Merged master, moved new config vals to v32

File: data/appRateLimiter.go [hunk 18]
```diff
@@ -0,0 +1,305 @@
...
+		if txgroup[i].Txn.Type == protocol.ApplicationCallTx {
+			appIdx := txgroup[i].Txn.ApplicationID
+			// hash appIdx into a bucket, do not use modulo without hashing first since it could
+			// assign two vanilla (and presumable, popular) apps to the same bucket.
+			keysBuckets.buckets = append(keysBuckets.buckets, txnToBucket(appIdx))
+			keysBuckets.keys = append(keysBuckets.keys, txnToDigest(appIdx))
+			if len(txgroup[i].Txn.ForeignApps) > 0 {
+				for _, appIdx := range txgroup[i].Txn.ForeignApps {
+					keysBuckets.buckets = append(keysBuckets.buckets, txnToBucket(appIdx))
+					keysBuckets.keys = append(keysBuckets.keys, txnToDigest(appIdx))
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply data/appRateLimiter.go:275]: Foreign app is the only way to assume/expect app call, and this might be under estimation to count it once (like a single foreign app can be called multiple times).

File: data/txHandler.go [hunk 19]
```diff
@@ -179,13 +182,26 @@ func MakeTxHandler(opts TxHandlerOpts) (*TxHandler, error) {
 	}
 
 	if opts.Config.EnableTxBacklogRateLimiting {
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply data/txHandler.go]: added a new config option for this: `EnableAppTxBacklogRateLimiting`

File: data/appRateLimiter.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (jasonpaulos) [reply data/appRateLimiter.go:275]: After more thought, I think it could make more sense to only count unique apps that appear in a transaction group. Here's my reasoning:
* The situation we want to avoid is people flooding the network with many calls to a single app which will race -- only a very small amount of the app calls will be accepted, while the vast majority will either be rejected or invalid for some other reason.
* App calls in the same transaction group, as well as inner transactions, are all atomically linked. In a flooding/race situation, there's no advantage to packing a larger group or issuing more inners, since if any transaction fails, the entire group fails.
* If we want to throttle apps which are experiencing flood/race events it's therefore sufficient to measure only the unique apps in an atomic group -- whether that group calls the same app 1 time or 16, it shouldn't make a difference. By not measure the unique apps, we are giving additional weight to some types of app interactions which don't necessarily deserve to be throttled to a higher extent.

If you agree, great, but if not, I still think this PR is beneficial as-is, so I'm still prepared to approve (just looking through most recent changes now).

**Review by jasonpaulos — APPROVED** [review]

File: data/appRateLimiter.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (algorandskiy) [reply data/appRateLimiter.go:275]: not counting foreign apps per txn in a group would allow 16x more calls per some app repeatedly mentioned in a group.

File: data/appRateLimiter.go [hunk 20]
```diff
@@ -0,0 +1,305 @@
...
+		var key keyType
+		copy(key[:], h[:len(key)])
+		return key
+	}
+	txnToBucket := func(appIdx basics.AppIndex) int {
+		return int(memhash64(uint64(appIdx), seed) % uint64(numBuckets))
+	}
+	for i := range txgroup {
+		if txgroup[i].Txn.Type == protocol.ApplicationCallTx {
+			appIdx := txgroup[i].Txn.ApplicationID
```
Comment (jasonpaulos) [inline data/appRateLimiter.go:267]: I just thought of this, if this is an app create call, `appIdx` will be 0. There should probably be a special case to ignore this situation, otherwise we could be rate limiting app creates.

I'm not sure how widely used this is currently, but there does exist a pattern where you can create a new app with it's OnCompletion set to delete to have a temporary app -- this can be used to add budget to your group or perform some additional checks. I wouldn't want that use case to be affected by this.

File: config/localTemplate.go [hunk 21]
```diff
@@ -231,7 +231,20 @@ type Local struct {
...
 
-	// EnableTxBacklogRateLimiting controls if a rate limiter and congestion manager shouild be attached to the tx backlog enqueue process
+	// TxBacklogAppTxRateLimiterMaxSize denotes a max size for the tx rate limiter
+	// calculated as "a thousand apps on a network of thousand of peers"
+	TxBacklogAppTxRateLimiterMaxSize int `version[32]:"1048576"`
+
+	// TxBacklogAppTxPerSecondRate determines a target app per second rate for the app tx rate limiter
+	TxBacklogAppTxPerSecondRate int `version[32]:"100"`
+
+	// TxBacklogRateLimitingCongestionRatio determines the backlog filling threashold in percents at which app limiter kicks in
```
Comment (gmalouf) [inline config/localTemplate.go:241]:
```suggestion
	// TxBacklogRateLimitingCongestionRatio determines the backlog filling threshold percentage at which the app  tx rate limiter kicks in
```

File: config/localTemplate.go [hunk 22]
```diff
@@ -231,7 +231,20 @@ type Local struct {
 	// TxBacklogReservedCapacityPerPeer determines how much dedicated serving capacity the TxBacklog gives each peer
 	TxBacklogReservedCapacityPerPeer int `version[27]:"20"`
 
-	// EnableTxBacklogRateLimiting controls if a rate limiter and congestion manager shouild be attached to the tx backlog enqueue process
+	// TxBacklogAppTxRateLimiterMaxSize denotes a max size for the tx rate limiter
+	// calculated as "a thousand apps on a network of thousand of peers"
+	TxBacklogAppTxRateLimiterMaxSize int `version[32]:"1048576"`
```
Comment (gmalouf) [inline config/localTemplate.go:236]: For naming consistency with the overall knob, should these related settings all start with `AppTxBacklog` as their prefix? Make it easy to distinguish the TxRateLimiter settings from the the app tx rate limiter

File: data/appRateLimiter.go (hunk 20 above)
Comment (algorandskiy) [reply data/appRateLimiter.go:278]: done

File: data/appRateLimiter.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (algorandskiy) [reply data/appRateLimiter.go]: implemented

File: config/localTemplate.go (hunk 22 above)
Comment (algorandskiy) [reply config/localTemplate.go:236]: idk, TxBacklogXXX vs TxBacklogAppXXX looks consistent

File: config/localTemplate.go (hunk 21 above)
Comment (algorandskiy) [reply config/localTemplate.go:241]: applied

**Review by jasonpaulos — APPROVED** [review]
LGTM, thanks for incorporating the app ID changes

File: data/txHandler.go [hunk 23]
```diff
@@ -178,14 +181,29 @@ func MakeTxHandler(opts TxHandlerOpts) (*TxHandler, error) {
...
-	if opts.Config.EnableTxBacklogRateLimiting {
-		rateLimiter := util.NewElasticRateLimiter(
-			txBacklogSize,
-			opts.Config.TxBacklogReservedCapacityPerPeer,
-			time.Duration(opts.Config.TxBacklogServiceRateWindowSeconds)*time.Second,
-			txBacklogDroppedCongestionManagement,
-		)
-		handler.erl = rateLimiter
+	if opts.Config.EnableTxBacklogRateLimiting || opts.Config.EnableTxBacklogAppRateLimiting {
+		if opts.Config.TxBacklogRateLimitingCongestionPct > 100 || opts.Config.TxBacklogRateLimitingCongestionPct < 0 {
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply data/txHandler.go:185]: TxBacklogRateLimitingCongestionPct is used for both limiters

**Review by gmalouf — APPROVED** [review]
